delta-io/connectors
# Flink Delta Source PR_9 - Builder

This proposition for Delta source builder provides dedicated factory methods that will create instance of Delta source builder for `RowData` type, each for one of the boundedness modes. The methods are:
`BoundedDeltaSourceBuilder<RowData>` and `ContinuousDeltaSourceBuilder<RowData>`

The API is:
```
DeltaSource<RowData> source = DeltaSource.forBoundedRowData(
                new Path("s3://some/path"),
                COLUMN_NAMES,
                COLUMN_TYPES,
                hadoopConf
            )
            .versionAsOf(10)
            .build();
```
and

```
DeltaSource<RowData> source = DeltaSource.forContinuousRowData(
                new Path("s3://some/path"),
                COLUMN_NAMES,
                COLUMN_TYPES,
                hadoopConf
            )
            .updateCheckIntervalMillis(1000)
            .startingVersion(10)
            .build();
```
More examples can be found in test folder `io.delta.flink.source.SourceExamples` class

The creation of `RowData` format is hidden from the end user. 
When in future we would like to provide new formats for example POJO or String format (so the source can produce elements with type other than RowType) we decided that we will expose format to the end user.

-------

Builder for DeltaSource.
Added `DeltaSourceInternal` and move it to internal package to hide all Flink's API methods from Javadoc.
"Public" `DeltaSource` class has only static methods to create builders.

Skipping PR 8 for now.
Examples for both builders are added to `SourceExamples` class in test folder. 

PR Plan:
PR 5 - Option support in Bounded Mode
PR 6 - Continuous mode without monitoring for changes
PR 6.1 - Package Refactoring
PR 7 - Monitoring for changes in Continuous mode
PR 7.1 More tests for PR 7 including IT case tests for Continuous mode.
PR 8 - Partition Support (POSTPONED)
**PR 9 - Builder**
PR 9.1 - Add validation for option names and values including using not applicable options for given mode via generic option method.

"Begin with the end in mind" -> aim to release first week or so of June, e.g. June 7
Priorities this week
PR 9.1 - builder API (extra validation to come in another PR, after PR 8)
PR 8 - partitioning support -> Initial PR ready for May 4th.

While @kristoffSC is on vacation
@tdas + @scottsand-db  to review public API
Once Krzysztof is back
README / docs
QA
end-to-end integration tests - using Sink to write, Source to read, etc.
more examples / migrate examples from internal flink module to examples folder

TODOs and other improvements planed as next:

- Use index for keeping track of already processed paths in SnapshotProcessor - needs new Delta API.
- Do do not create snapshot when streaming only changes (`startingVersion` or `startingTimestamp` option used). - needs new Delta API since for startingTimestamp we still need to create a snapshot. There is no API like deltaLog.getChangesFromTimestamp(....)
- Refactor Test util classes to make them common for sink and source.
- Re-visit Delta Standalone APIs for when the source see's metadata/protocol actions. Update ActionProcessor class to handle MetaData and Protocol Actions
